### PR TITLE
Allow multiple endpoints with DMA capability

### PR DESCRIPTION
**Pull request by zhemao**

### CODE REVIEW

File: src/main/scala/junctions/ReorderQueue.scala [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  val valid = Bool(INPUT)
+  val tag = UInt(INPUT, tagWidth)
+  val data = dType.cloneType.asOutput
+  val matches = Bool(OUTPUT)
+
+  override def cloneType =
+    new ReorderDequeueIO(dType, tagWidth).asInstanceOf[this.type]
+}
+
+class ReorderQueue[T <: Data](dType: T, tagWidth: Int,
```
Comment (davidbiancolin) [inline src/main/scala/junctions/ReorderQueue.scala]: Does this not exist in RC?

File: src/main/scala/midas/core/FPGATop.scala [hunk 2]
```diff
@@ -10,6 +10,7 @@ import chisel3.util._
 import chisel3.core.ActualDirection
 import chisel3.core.DataMirror.directionOf
 import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.diplomacy.AddressSet
```
Comment (davidbiancolin) [inline src/main/scala/midas/core/FPGATop.scala]: This is mildly blasphemous. :P

File: src/main/scala/midas/core/FPGATop.scala [hunk 3]
```diff
@@ -166,11 +172,16 @@ class FPGATop(simIoType: SimWrapperIO)(implicit p: Parameters) extends Module wi
     }
   }
 
-  assert(dmaPorts.size <= 1)
-  if (dmaPorts.nonEmpty) {
+  if (dmaPorts.isEmpty) {
+    io.dma := DontCare
```
Comment (davidbiancolin) [inline src/main/scala/midas/core/FPGATop.scala]: Should you not tie off the appropriate ready and valids instead, or is this safe?

**Comment by sagark** [comment]
Quick! Add a bunch of whitespace that we can delete later! :P

File: src/main/scala/junctions/ReorderQueue.scala (hunk 1 above)
Comment (zhemao) [reply src/main/scala/junctions/ReorderQueue.scala]: The version in rocket-chip doesn't have multiple dequeue interfaces, which is needed for the Nasti code.

File: src/main/scala/midas/core/FPGATop.scala (hunk 3 above)
Comment (zhemao) [reply src/main/scala/midas/core/FPGATop.scala]: I probably should tie them off, yeah.

**Pull request merged**